source/agora/common/BitMask.d

```diff
@@ -0,0 +1,192 @@
+{
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.range;
+
+    @safe:
+
+    /// Serialization
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        formattedWrite(sink, "%s", iota(this.length).map!(i => this[i] ? "1" : "0").join(""));
+    }
+
+    /// Support for Vibe.d serialization
+    public string toString () const
+    {
+        string ret;
+        scope void delegate (scope const char[]) @safe sink = (scope v) { ret ~= v; };
+        this.toString(sink);
+        return ret;
+    }
+
+    pure:
+
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto bitmask = BitMask(str.length);
+        str.enumerate.each!((i, c)
+        {
+            if (c == '1')
+                bitmask[i] = true;
+        });
+        return bitmask;
+    }
+
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t length;
+
+    /// Bytes to hold bits to indicate if a validator has signed
+    private ubyte[] bytes;
+
+    public this (size_t length) inout
+    {
+        this.length = length;
+        if (length > 0)
+            this.bytes = new inout(ubyte)[1 + ((length - 1) / 8)];
+    }
+
+    public this (size_t length, in ubyte[] bytes)
+    {
+        this(length);
+        assert(this.bytes.length == bytes.length);
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
+    }
+
+    // set the bits that are set in given BitMask
+    public auto opOpAssign (string op : "|") (in BitMask add)
+    {
+        assert(this.length == add.length, "BitMask assignment must be for same bit length");
+        iota(this.length).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
+        return this;
+    }
+
+    /// return the indices of bits set
+    public auto setIndices () const
+    {
+        return iota(this.length).filter!(i => this[i]);
+    }
+
+    /// return the indices of bits not set
+    public auto notSetIndices () const
+    {
+        return iota(this.length).filter!(i => !this[i]);
+    }
+
+    /// return the length of set bits
+    public size_t setCount () const
+    {
+        return this.setIndices.count!(i => this[i]);
+    }
+
+    /// Support for sorting by count of set bits not value
+    public int opCmp (in typeof(this) rhs) const
+    {
+        assert(this.length == rhs.length, "Comparing different sized BitMasks is not valid");
+        return this.setCount < rhs.setCount ? -1 : 1;
+    }
+
+    @nogc:
+
+    // support setting a bit (asserts if trying to unset a bit)
+    public auto opIndexAssign(bool set, size_t bit_index)
+    {
+        if (bit_index >= this.length)
+                assert(0, "Attempt to set index beyond length of bitmask");
+        const size_t byte_index = (bit_index) / 8;
+        const size_t bit_index_in_byte = bit_index % 8;
```

Comment:
Operation % is already implemented in function mask().

Reply:
Good catch, I missed that.